[CPU] Clean up primitive desc creation in prepareParams

EgorDuplensky opened the pull request:
Small refactoring

EgorDuplensky commented:
@maxnick Could you please take a look

EgorDuplensky commented on src/plugins/intel_cpu/src/nodes/softmax.cpp:
```diff
@@ -182,7 +181,7 @@ void SoftMax::prepareParams() {
+        primitive_desc_iterator itpd = prim_desc;
 
         while (itpd) {
             impl_desc_type impl_type = parse_impl_name(itpd.impl_info_str());
```
@zhangYiIntel May I ask you to take a look at this?
Will it be correct to re-design this logic the same way it is done for matmul, i.e. just pick first available implementation in case key.implType == ref_any?

EgorDuplensky replied:
Yeah, that is why I have not touched it. Lets leave it for now.

EgorDuplensky commented on src/plugins/intel_cpu/src/nodes/pooling.cpp:
```diff
@@ -370,28 +370,23 @@ void Pooling::prepareParams() {
-                prim_desc = itpd.get();
-                break;
-            }
-            if (!itpd.next_impl())
-                break;
```
It seems pooling primitive has a different way of handling the case when a proper implementation is not found.
It just uses the last implementation (reference), when other primitive either return nullptr or pick the first available implementation.
@dmitry-gorokhov @maxnick Does anyone know the reason. Can we pick the first available instead?

EgorDuplensky replied:
I have changed the behavior to use the first available implementation and at least the tests which cover the case with the fallback are passed.

maxnick replied:
Actually no. On L384 the first available implementation is stored, then if nothing matched the impl_type, the stored descriptor (witch corresponds to the first available implementation) is used.

maxnick commented:
> @maxnick Could you please take a look

The intention of this PR is understandable. But, the thing is little bit more complicated. We use this while loop to match the implementation selected on the compilation stage. We do this intentionally to avoid possible extra reordering. On the compilation stage the variable data layout is selected and propagated through the whole graph in a way to avoid reordering as much as possible. The oneDNN implementations types are stored to ensure the same data layout at runtime. Now, if we always select the first available implementation, we cannot be sure that the layout is the same as stored in the node descriptor, which means that there is a probability of the layout mismatch and insertion of additional reorders, or even exception throwing, if the layout mismatch handling is not implemented for some node. Yes, it is reasonable to assume that in the most cases the layout will be the same, and if not, it is handled with the implicit reordering logic anyway, but it is not obvious what is better to do in case when the first implementation does not match the selected, but the selected still persists in the implementations list.
So the question is, are we sure that this refactoring is at least safe? I mean is it possible that the first implementation always matches the layout assigned to the edge tensor descriptor.

EgorDuplensky commented:
> > @maxnick Could you please take a look
> 
> The intention of this PR is understandable. But, the thing is little bit more complicated. We use this while loop to match the implementation selected on the compilation stage. We do this intentionally to avoid possible extra reordering. On the compilation stage the variable data layout is selected and propagated through the whole graph in a way to avoid reordering as much as possible. The oneDNN implementations types are stored to ensure the same data layout at runtime. Now, if we always select the first available implementation, we cannot be sure that the layout is the same as stored in the node descriptor, which means that there is a probability of the layout mismatch and insertion of additional reorders, or even exception throwing, if the layout mismatch handling is not implemented for some node. Yes, it is reasonable to assume that in the most cases the layout will be the same, and if not, it is handled with the implicit reordering logic anyway, but it is not obvious what is better to do in case when the first implementation does not match the selected, but the selected still persists in the implementations list. So the question is, are we sure that this refactoring is at least safe? I mean is it possible that the first implementation always matches the layout assigned to the edge tensor descriptor.

Actually there is no code logic changed, just cleaned up code duplications and more declarative style (I have reverted the deep copy change, which would introduce a perf degradation but also did not affect the overall code logic).
May I ask you point out the piece of code you have worries about, maybe I have missed something?

maxnick commented on src/plugins/intel_cpu/src/nodes/matmul.cpp:
```diff
@@ -594,46 +594,36 @@ void MatMul::prepareParams() {
 
-        primitive_desc_iterator itpd = matmul_desc;
-        matmul::primitive_desc prim_desc;
+        auto initial_desc = dnnl::matmul::primitive_desc(prim_desc.get());
```
```suggestion
        auto first_desc = dnnl::matmul::primitive_desc(prim_desc.get());
```

maxnick reviewed (approved):
In general LGTM. There are a couple of minor comments left.

maxnick commented:
> > > @maxnick Could you please take a look
> > 
> > 
> > The intention of this PR is understandable. But, the thing is little bit more complicated. We use this while loop to match the implementation selected on the compilation stage. We do this intentionally to avoid possible extra reordering. On the compilation stage the variable data layout is selected and propagated through the whole graph in a way to avoid reordering as much as possible. The oneDNN implementations types are stored to ensure the same data layout at runtime. Now, if we always select the first available implementation, we cannot be sure that the layout is the same as stored in the node descriptor, which means that there is a probability of the layout mismatch and insertion of additional reorders, or even exception throwing, if the layout mismatch handling is not implemented for some node. Yes, it is reasonable to assume that in the most cases the layout will be the same, and if not, it is handled with the implicit reordering logic anyway, but it is not obvious what is better to do in case when the first implementation does not match the selected, but the selected still persists in the implementations list. So the question is, are we sure that this refactoring is at least safe? I mean is it possible that the first implementation always matches the layout assigned to the edge tensor descriptor.
> 
> Actually there is no code logic changed, just cleaned up code duplications and more declarative style used (I have reverted the deep copy change, which would introduce a perf degradation but also did not affect the overall code logic). May I ask you to point out the piece of code you have worries about, maybe I have missed something?

Yep, got the idea. You simply extracted the while loops to a separate method.

EgorDuplensky commented:
@maxnick I guess we can merge it after a rebase 👀

EgorDuplensky commented on src/plugins/intel_cpu/src/nodes/pooling.cpp:
```diff
@@ -370,30 +370,29 @@ void Pooling::prepareParams() {
+
+        dnnl::pooling_forward::primitive_desc tmp = prim_desc.get();
+
+        auto initial_desc = dnnl::pooling_forward::primitive_desc(prim_desc.get());
```
Done
